Various fixes regarding PKCS#12 input and related cleanup of apps, doc, and tests

DDvO opened the pull request:
***UPDATE:*** Originally this PR was about extending support for PKCS#12 input in apps.
I've meanwhile carved out the most interesting pieces of that and contributed them separately.
This is the leftovers fixing several corner cases in PKCS#12 input and its error handling.
There are also some rather unrelated fixes to several apps and their documentation, which I could separate if requested.     

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

DDvO commented:
This PR also contains (in its first commit) a minor fix of a conditionally unused variable,
which caused a build failure when OPENSSL_SYS_UNIX is not defined and strict/pedantic compiler settings (-Werror) are used.

DDvO commented:
I also noticed that doc/man1/s_server.pod is somewhat out of sync.
Besides my adaptations that are related to the PKCS#12 extensions I made,
I removed the following references to options not available any more:
```
[B<-xkey>]
[B<-xcert>]
[B<-xchain>]
[B<-xchain_build>]
[B<-xcertform PEM|DER>]
[B<-xkeyform PEM|DER>]
```

bernd-edlinger commented:
I think if you have a PKCS#12 file, then you have the certificate, private key and any chain
certificates all in one place, so you should only name _one_ pkcs#12 file, and one password
and not have three different file paths, but only one password.

DDvO commented:
I agree that typically a PKCS#12 file contains these related types of key material and that it would be logical and most convenient to refer to such a file only once (including any password input).
Yet so far the OpenSSL CLI, except for the pkcs12 app, does not support this.
And I just noticed that, for some reason, even pkcs12.c does not make use of load_pkacs12() to load them jointly.

I fear that rectifying this would take some non-negligible effort, including a (backward compatible) extension of the CLI options design for all apps that should support joint PKCS#12 input. Shall we go for that, and who would be willing to help doing this?

Nevertheless, the generaliizations I've proprosed here are already useful in themselves - with just the inconvenience that, as before, a PKCS#12 file used not only for key input but also for certificate input needs to be named (together with any password input) more than once.

FdaSilvaYY commented on apps/ca.c:
```diff
@@ -171,10 +172,11 @@ const OPTIONS ca_options[] = {
-    {"key", OPT_KEY, 's', "Key to decode the private key if it is encrypted"},
+    {"key", OPT_KEY, 's', "Key to decode the private key or cert files if encrypted. Better use -passin"},
     {"cert", OPT_CERT, '<', "The CA cert"},
+    {"certform", OPT_CERTFORM, 'f', "Certifiate file format (PEM|DER|P12)"},
```
s/certifiate/certificate

DDvO replied:
Oops, my favorite typo - done.

bernd-edlinger commented on apps/s_client.c:
```diff
@@ -620,13 +620,13 @@ const OPTIONS s_client_options[] = {
     {"key", OPT_KEY, 's', "Private key file to use, if not in -cert file"},
-    {"keyform", OPT_KEYFORM, 'E', "Key format (PEM, DER or engine) PEM default"},
-    {"pass", OPT_PASS, 's', "Private key file pass phrase source"},
+    {"keyform", OPT_KEYFORM, 'E', "Key format (PEM, DER, P12, or engine) PEM default"},
```
and, please remove the comma before "or"

kaduk replied:
The comma before "or" seems like a perfectly valid serial comma usage to me.
I don't see a need to revisit the capitalization of "engine" here as part of this change.

DDvO replied:
So I leave this as it is.
On the comma topic: https://en.wikipedia.org/wiki/Serial_comma

bernd-edlinger commented on doc/man1/ca.pod:
```diff
@@ -30,10 +30,11 @@ B<openssl> B<ca>
 [B<-policy arg>]
 [B<-keyfile arg>]
-[B<-keyform PEM|DER>]
+[B<-keyform PEM|DER|ENGINE>]
```
I would very much prefer something like an option like -pkcs12 <file>,
that replaces all key and certificate options (is usable as an alternative).
Do you agree?

DDvO replied:
Good idea to have such an additional way for specifiying credential input in PKCS#12 format.
I've just added `-pkcs12in <file>` with 'pkcs12in' rather than 'pkcs12' to prevent potential (future) ambiguities, 
which implies `-keyfile <file> -cert <file> -keyform P12 -certform P12`.

I also updated `ca.pod` according to both above comments.
Doing so, I rearranged some of the explanations there for consistency and improved readability.

DDvO commented on apps/apps.c:
```diff
@@ -959,7 +965,32 @@ void* app_malloc(int sz, const char *what)
 {
-    return load_certs_crls(file, format, pass, desc, certs, NULL);
+    int ret = 0;
+    if (format == FORMAT_PKCS12) {
```
@FdaSilvaYY, thanks for your careful review.

Done, also after `BIO *bio = `... and after `X509 *cert = `... .

DDvO commented on apps/s_server.c:
```diff
@@ -786,22 +786,22 @@ const OPTIONS s_server_options[] = {
     {"keyform", OPT_KEYFORM, 'f',
      "Key format (PEM, DER or ENGINE) PEM default"},
-    {"pass", OPT_PASS, 's', "Private key file pass phrase source"},
+    {"pass", OPT_PASS, 's', "Private key and certifiate file pass phrase source"},
```
Done.

DDvO commented on doc/man1/s_server.pod:
```diff
@@ -264,22 +258,22 @@ followed by "length" bytes of extension data).  If the client sends
 =item B<-pass val>
 
-The private key password source. For more information about the format of B<val>
+The private key and certifiate file password source. For more information about the format of B<val>
```
Done.

DDvO commented on apps/apps.c:
```diff
@@ -649,10 +653,12 @@ X509 *load_cert(const char *file, int format, const char *cert_descrip)
     } else if (format == FORMAT_PEM) {
         x = PEM_read_bio_X509_AUX(cert, NULL,
-                                  (pem_password_cb *)password_callback, NULL);
+                                  (pem_password_cb *)password_callback, &cb_data);
```
I also see little point doing so for a cert-only file, but 
* a PEM file might contain also other structures including private keys, and it looks like in such case the while file can be encrypted (similarly to PKCS#12 files) and
* `PEM_read_bio_X509_AUX()` does support a password callback argument, 

so at least it should not hurt supplying it just in case.

DDvO commented on apps/apps.c:
```diff
@@ -649,10 +653,12 @@ X509 *load_cert(const char *file, int format, const char *cert_descrip)
     } else if (format == FORMAT_PKCS12) {
-        if (!load_pkcs12(cert, cert_descrip, NULL, NULL, NULL, &x, NULL))
-            goto end;
+        EVP_PKEY *pkey = NULL; /* &pkey is required by PKCS12_parse */
```
I had passed in a valid `&pkey` because according to the documentation of `PKCS12_pase()` this is required. Yet it turns out that this is not really the case.
So I've corrected the documentation, removed the needless `pkey` variable, and fixed some style nits.

DDvO replied:
I've kept the declaration of `cb_data` where it was because, as written above,
I believe there can be situations where passing in the password while parsing a cert from a PEM file can make sense.

DDvO replied:
It turned out that still a fix of `PKCS12_parse()` in case `pkey == NULL` and `cert != NULL` was needed, but now everything works fine.

DDvO commented:
> I had these lying around... can't even remember when I wrote them.

Thanks for these comments. I've just handled them.

DDvO commented:
The Travis red cross is due to the (unrelated) issue handled in #11573.

@levitte, all comments have been handled.

DDvO commented:
The two CI failures currently reported here are unrelated.

Ready for further reviewing.
See in particular the fixes to `PKCS12_parse()` and its documentation.

DDvO commented:
I've rebased this and partly squashed the commits.
@levitte, is there anything else to improve in this PR?

Hopefully this can soon be merged for inclusion in #11470.

mattcaswell commented:
Ping @levitte

FdaSilvaYY commented on apps/ca.c:
```diff
@@ -214,17 +213,21 @@ const OPTIONS ca_options[] = {
-    {"certform", OPT_CERTFORM, 'F',
-     "certificate input format (DER or PEM); default PEM"},
+    {"certform", OPT_CERTFORM, 'f',
+     "Certifiate file format (PEM|DER|P12|HTTP); default PEM"},
```
Typo : _Certifiate_

DDvO replied:
Ah, you caught my most notorious typo :sweat_smile:

FdaSilvaYY commented on apps/lib/apps.c:
```diff
@@ -831,7 +828,35 @@ void* app_malloc(int sz, const char *what)
+        BIO *bio = bio_open_default(file, 'r', format);
+
+        if (bio != NULL) {
+            PW_CB_DATA cb_data;
```
nit: blank line after declaration

DDvO replied:
fixed

FdaSilvaYY commented on apps/lib/apps.c:
```diff
@@ -831,7 +828,35 @@ void* app_malloc(int sz, const char *what)
+    } else if (format == FORMAT_ASN1) { /* load only one cert in this case */
+        X509 *cert = load_cert_pass(file, format, pass, desc);
+
+        if (cert) {
```
nit: `if (cert !=NULL)`

DDvO replied:
fixed

DDvO commented:
Thanks @FdaSilvaYY for having a look.
Rebased this PR and fixed all reported nits.

DDvO commented:
Travis failure is unrelated (No space left on device).

DDvO commented:
Travis CI failure is unrelated:
```
./INSTALL.md:1473: MD009 Trailing spaces
```
@levitte, please fixup your commit b19b983017f3865b1b3411a4e635a670d5798774.

DDvO commented:
Rebased to include fix #13097 of the `INSTALL.md` whitespace issue.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

levitte commented on apps/ca.c:
```diff
@@ -1364,14 +1365,12 @@ static int certify(X509 **xret, const char *infile, int informat,
     if (i < 0) {
         ok = 0;
         BIO_printf(bio_err, "Signature verification problems....\n");
-        ERR_print_errors(bio_err);
```
If you're going to replace this with `goto end` everywhere, it would be nice if you added back this call at the `end:` label.

DDvO replied:
Right - I had thought that `ERR_print_errors(bio_err);` will be called anyway in such cases by `ca_main()`,
but in case `ok` >= 0 this is not really the case. 
So I (re-)added  `ERR_print_errors(bio_err);` at the end of `certify()`.

DDvO replied:
On this occasion I've just done some further code simplifications in the body of `certify()`.

levitte commented on apps/pkcs12.c:
```diff
@@ -596,7 +599,8 @@ int pkcs12_main(int argc, char **argv)
         if (p12 == NULL) {
-            ERR_print_errors(bio_err);
+            BIO_printf(bio_err, "Error creating PKCS12 structure for %s\n",
+                       outfile);
```
If you look anywhere else, you'll see that we do `ERR_print_errors()` *in addition* to whatever the app wants to print.  The app error is not a replacement, so while I see nothing wrong with the `BIO_printf()` call, I do think it's wrong to remove the `ERR_print_errors()` call.  The latter is likely to give additional details.

DDvO replied:
That's understood.
Note that I did not remove `ERR_print_errors(bio_err);` but moved it after the `export_end:` label within `pkcs12_main()`
because outputting the error queue may be helpful also other (error) cases that end up there.

levitte commented on apps/storeutl.c:
```diff
@@ -458,18 +460,37 @@ static int process(const char *uri, const UI_METHOD *uimeth, PW_CB_DATA *uidata,
+                if (pkey == NULL) {
+                    items--;
+                    BIO_printf(bio_err, "Key not actually present\n");
+                    ret++;
```
Uhmmmmmmmm...  when does this happen?

DDvO replied:
I cannot recall in detail (I added this case exactly half a year ago) in a commit with the message
```
    apps/storeutl: Fix in case PKCS#12 input has no key or no user cert
```
but I think this may happen if for some reason the PKCS#12 input has a cert but not a key.
At least this defensive check should not hurt.

DDvO replied:
I've just tried this out with a PKCS#12 file not having a key.
It looks like in `case OSSL_STORE_INFO_PKEY:` this cannot happen, so I removed that check again.

levitte commented on apps/storeutl.c:
```diff
@@ -458,18 +460,37 @@ static int process(const char *uri, const UI_METHOD *uimeth, PW_CB_DATA *uidata,
+                if (cert == NULL) {
+                    items--;
+                    BIO_printf(bio_err, "Certificate not actually present\n");
+                    ret++;
```
Again, when does this happen?

DDvO replied:
Analogously here:
I cannot recall in detail (I added this case exactly half a year ago) in a commit with the message
```
    apps/storeutl: Fix in case PKCS#12 input has no key or no user cert
```
but I think this may happen if for some reason the PKCS#12 input has a key but not a cert.
At least this defensive check should not hurt.

DDvO replied:
I've just tried this out with a PKCS#12 file not having a cert.
It looks like in `case OSSL_STORE_INFO_CERT:` this cannot happen, so I removed that check again.

levitte commented on test/recipes/80-test_pkcs12.t:
```diff
@@ -66,17 +66,19 @@ ok(run(test(["pkcs12_format_test"])), "test pkcs12 formats");
-my $tmpfile = "tmp.p12";
+my $outfile1 = "out1.p12";
+my $outfile2 = "out2.p12";
+my $outfile3 = "out3.p12";
```
Good find!

DDvO commented:
Thanks @levitte for having a look again at this PR!
I've just addressed all new comments.

Would be glad if we can finalize this PR before its 3th anniversary in December :wink:

t8m commented:
Just FYI @DDvO  - post 3.0.0 doesn't mean it cannot be merged for 3.0.0.

t8m commented:
It's just it isn't blocking 3.0.0

DDvO commented:
> Just FYI @DDvO - post 3.0.0 doesn't mean it cannot be merged for 3.0.0.
> It's just it isn't blocking 3.0.0

Thanks for pointing this out!

Would be nice if this PR still goes into 3.0.0. since it fixes a couple of corner cases and improves documentation.

Since this PR has already been handled pretty far it should not make much (more) effort to finish it up very soon.
Conversely, long delays in handling PRs actually causes more effort in total 
because everyone involved needs to mentally dig into the same contents again and again
and there is also much higher changes for merge conflicts to build up causing extra work.

DDvO commented:
@levitte, I addressed all comments you gave two days back and did respective changes.
Is there anything else to improve?

DDvO commented:
Ping @levitte

DDvO commented:
Is there anything more to improve on this PR?

t8m commented:
Removed the Post 3.0.0 milestone as we decided to use the milestone to mark things that should not go in to 3.0.0.

t8m commented:
@DDvO why the addition of the two .p12 files? I do not see them being used.

t8m commented:
Except for the question above, this LGTM.

DDvO commented:
> @DDvO why the addition of the two .p12 files? I do not see them being used.

Ah, good point - the files `test/v3-certs{,-descert}.p12` where leftovers superseded by `test/certs/v3-certs-{RC2,TDES}.p12`.
So I've removed them (and rebased the PR).

> Except for the question above, this LGTM.

Pleased to hear!

openssl-machine commented:
This pull request is ready to merge

DDvO commented:
Merged :sweat_smile: 

I'm glad that this nearly three years old PR, after various adaptations and many long waiting times in between, finally has been brought to a good end, even in time for inclusion in 3.0.

Thanks again to @levitte and to all other reviewers for their comments and to @t8m for providing the final push.
